common/route/router2.cc

```diff
@@ -559,25 +563,27 @@ struct Router2
     }
 
     // Functions for marking wires as visited, and checking if they have already been visited
-    void set_visited_fwd(ThreadContext &t, int wire, PipId pip)
+    void set_visited_fwd(ThreadContext &t, int wire, PipId pip, float cost)
     {
         auto &wd = flat_wires.at(wire);
         if (!wd.visited_fwd && !wd.visited_bwd)
             t.dirty_wires.push_back(wire);
         wd.pip_fwd = pip;
         wd.visited_fwd = true;
+        wd.cost_fwd = cost;
     }
-    void set_visited_bwd(ThreadContext &t, int wire, PipId pip)
+    void set_visited_bwd(ThreadContext &t, int wire, PipId pip, float cost)
     {
         auto &wd = flat_wires.at(wire);
         if (!wd.visited_fwd && !wd.visited_bwd)
             t.dirty_wires.push_back(wire);
         wd.pip_bwd = pip;
         wd.visited_bwd = true;
+        wd.cost_bwd = cost;
     }
 
-    bool was_visited_fwd(int wire) { return flat_wires.at(wire).visited_fwd; }
-    bool was_visited_bwd(int wire) { return flat_wires.at(wire).visited_bwd; }
+    bool was_visited_fwd(int wire, float cost) { return flat_wires.at(wire).visited_fwd && flat_wires.at(wire).cost_fwd <= cost; }
```

Comment:
at some point it might be worth looking at adding an epsilon like router1 to avoid minimally-useful revisits, but that's probably a future exercise